go/chat/boxer.go

```diff
@@ -402,8 +412,276 @@ func (b *Boxer) unboxV1(ctx context.Context, boxed chat1.MessageBoxed, encryptio
+	}, nil
+}
+
+// Unversions a header.
+// Also check that the HeaderSignature field from MessageBoxed V1 is nil.
+// Therefore only for use with MessageBoxed V2.
+// Returns (header, bodyHash, err)
+func (b *Boxer) unversionHeader(ctx context.Context, headerVersioned chat1.HeaderPlaintext) (chat1.MessageClientHeaderVerified, []byte, UnboxingError) {
+	headerVersion, err := headerVersioned.Version()
+	if err != nil {
+		return chat1.MessageClientHeaderVerified{}, nil, NewPermanentUnboxingError(err)
+	}
+	switch headerVersion {
+	case chat1.HeaderPlaintextVersion_V1:
+		hp := headerVersioned.V1()
+		if hp.HeaderSignature != nil {
+			return chat1.MessageClientHeaderVerified{}, nil,
+				NewPermanentUnboxingError(fmt.Errorf("HeaderSignature non-nil in MBV2"))
+		}
+		return chat1.MessageClientHeaderVerified{
+			Conv:         hp.Conv,
+			TlfName:      hp.TlfName,
+			TlfPublic:    hp.TlfPublic,
+			MessageType:  hp.MessageType,
+			Prev:         hp.Prev,
+			Sender:       hp.Sender,
+			SenderDevice: hp.SenderDevice,
+			// CORE-4540: MerkleRoot will be in signed header.
+			OutboxID:   hp.OutboxID,
+			OutboxInfo: hp.OutboxInfo,
+		}, hp.BodyHash, nil
+	default:
+		return chat1.MessageClientHeaderVerified{}, nil,
+			NewPermanentUnboxingError(NewHeaderVersionError(headerVersion,
+				b.headerUnsupported(ctx, headerVersion, headerVersioned)))
+	}
+}
+
+func (b *Boxer) unversionBody(ctx context.Context, bodyVersioned chat1.BodyPlaintext) (chat1.MessageBody, UnboxingError) {
+	bodyVersion, err := bodyVersioned.Version()
+	if err != nil {
+		return chat1.MessageBody{}, NewPermanentUnboxingError(err)
+	}
+	switch bodyVersion {
+	case chat1.BodyPlaintextVersion_V1:
+		return bodyVersioned.V1().MessageBody, nil
+	default:
+		return chat1.MessageBody{},
+			NewPermanentUnboxingError(NewBodyVersionError(bodyVersion,
+				b.bodyUnsupported(ctx, bodyVersion, bodyVersioned)))
+	}
+}
+
+func (b *Boxer) verifyBodyHash(ctx context.Context, bodyEncrypted chat1.EncryptedData, bodyHashSigned []byte) UnboxingError {
+	bodyHashObserved := b.hashV1(bodyEncrypted.E)
+
+	if len(bodyHashSigned) == 0 {
+		return NewPermanentUnboxingError(BodyHashInvalid{})
+	}
+
+	if !libkb.SecureByteArrayEq(bodyHashObserved, bodyHashSigned) {
+		return NewPermanentUnboxingError(BodyHashInvalid{})
+	}
+	return nil
+}
+
+// Compare the unsigned and signed header for MessageBoxedVersion_V1.
+// The V1 and V2 checks are different methods because they are strict on slightly different things.
+// Confirm that fields in the server-supplied ClientHeader match what
+// we decrypt. It would be preferable if the server didn't supply this data
+// at all (so that we didn't have to worry about anyone trusting it
+// *before* we get to this check, for example), but since we have it we
+// need to check it.
+// The most important check here is that the Sender and SenderDevice match.
+// That is the only thing that gives the verification key used credibility.
```

Comment:
I hadn't thought of using the KID to look up the UID+DID. What you said makes sense. Here's another summary of what this does for v2.
The inputs are: `unsigned sender info`, `kid`, `signed sender info`. We check that the `kid` is valid for the `unsigned sender info` and then after opening the header we check that the `signed sender info` equals the `unsigned sender info`. Then we output `signed sender info` on `MessageClientHeaderVerified`.

I just noticed we don't check that the key is for the correct device. So I guess we should do that.